src/backend/workers/go.remoteIDs/jobs.go

```diff
@@ -1,88 +1,98 @@
 package go_remoteIDs
 
 import (
+	"crypto/sha256"
+	"errors"
+	"fmt"
 	. "github.com/CaliOpen/Caliopen/src/backend/defs/go-objects"
 	log "github.com/Sirupsen/logrus"
+	"sync"
 )
 
-// AddJobFor parses remote identity data to build appropriate job and adds it to MainCron
-func (p *Poller) AddJobFor(idkey string) (err error) {
-	p.cacheMux.Lock()
-	defer p.cacheMux.Unlock()
-	if entry, ok := p.Cache[idkey]; ok {
-		switch entry.remoteProtocol {
-		case EmailProtocol, ImapProtocol:
-			cronStr := "@every " + entry.pollInterval + "m"
-			entry.cronId, err = p.MainCron.AddJob(cronStr, imapJob{
-				remoteId:  entry.remoteID.String(),
-				natsTopic: p.Config.NatsTopics["imap"],
-				poller:    p,
-				userId:    entry.userID.String(),
-			})
-			if err != nil {
-				log.WithError(err).Warn("[AddJobFor] failed to add job to MainCron")
-				return
-			}
-			p.Cache[idkey] = entry
-		case TwitterProtocol:
-			cronStr := "@every " + entry.pollInterval + "m"
-			entry.cronId, err = p.MainCron.AddJob(cronStr, twitterJob{
-				remoteId:  entry.remoteID.String(),
-				natsTopic: p.Config.NatsTopics["twitter_dm"],
-				poller:    p,
-				userId:    entry.userID.String(),
-			})
-			if err != nil {
-				log.WithError(err).Warn("[AddJobFor] failed to add job to MainCron")
-				return
-			}
-			p.Cache[idkey] = entry
-			p.addWorkerFor(idkey)
-		default:
-			log.WithError(err).Warnf("[AddJobFor] unknow Remote Identity protocol <%s>", entry.remoteProtocol)
-			return
-		}
-		return
-	} else {
-		log.WithError(err).Warnf("[AddJobFor] failed to retrieve cache key <%s>", idkey)
-		return
+type JobsHandler struct {
+	pendingJobs  map[string]map[[32]byte]Job // pattern is [worker][job-hash]job
+	jobsSequence map[string][][32]byte       // pattern is [worker][]job-hash <- FIFO ordered
+	jobsMux      *sync.Mutex
+}
+
+type Job struct {
+	Worker string // protocol worker like email, twitter, etc. as defined in idpoller's config
+	Order  BrokerOrder
+}
+
+func initJobsHandler() (*JobsHandler, error) {
+	return &JobsHandler{
+		map[string]map[[32]byte]Job{},
+		map[string][][32]byte{},
+		&sync.Mutex{},
+	}, nil
+}
+
+// AddPendingJob adds a job to _pending_ map only if job's hash does not already exists for job's worker.
+// pending jobs are ordered in a FIFO list per each worker.
+func (jh *JobsHandler) AddPendingJob(job Job) {
+	jh.jobsMux.Lock()
+	key := sha256.Sum256([]byte(job.Order.UserId + job.Order.IdentityId + job.Order.Order))
+	if _, ok := jh.pendingJobs[job.Worker]; !ok {
+		jh.pendingJobs[job.Worker] = make(map[[32]byte]Job)
 	}
+	if _, ok := jh.pendingJobs[job.Worker][key]; !ok {
+		jh.pendingJobs[job.Worker][key] = job
+		jh.jobsSequence[job.Worker] = append(jh.jobsSequence[job.Worker], key)
+	}
+	log.Debugf("pending jobs list : %+v\n", jh.pendingJobs)
+	log.Debugf("jobs sequence list : %+v\n", jh.jobsSequence)
+	log.Infof("[jobsHandler] jobs for workers <%s> updated : %d job(s) pending", job.Worker, len(jh.jobsSequence[job.Worker]))
+	jh.jobsMux.Unlock()
```

Comment:
unlock using defer

Reply:
I think it's useless again because every calls between Lock and Unlock cannot fail.